### PR TITLE
Enabled UCR auto image gc for DC/OS.

### PR DESCRIPTION
**Pull request by Gilbert88**
## High-level description

 - Enabled UCR auto image gc for dc/os.
 - This PR builds on top of [pull/2165](https://github.com/dcos/dcos/pull/2165)([DCOS_OSS-1945](https://jira.mesosphere.com/browse/DCOS_OSS-1945)).

## Corresponding DC/OS tickets (obligatory)

These DC/OS JIRA ticket(s) must be updated (ideally closed) in the moment this PR lands:

  - [DCOS_OSS-1933](https://jira.mesosphere.com/browse/DCOS_OSS-1933) Enable UCR basic auto image GC for DC/OS.

## Related tickets (optional)

Other tickets related to this change:

  - [DCOS_OSS-1945](https://jira.mesosphere.com/browse/DCOS_OSS-1945) Bump mesos just before Beta freeze.

## Checklist for all PRs

  - [ ] Included a test which will fail if code is reverted but test is not. If there is no test please explain here:
  - [x] Read the [DC/OS contributing guidelines](https://github.com/dcos/dcos/blob/master/contributing.md)
  - [x] Followed relevant code rules [Rules for Packages and Systemd](https://github.com/dcos/dcos/tree/master/docs)

## Checklist for component/package updates:

If you are changing components or packages in DC/OS (e.g. you are bumping the sha or ref of anything underneath `packages`), then in addition to the above please also include:

  - [ ] Change log from the last version integrated (this should be a link to commits for easy verification and review): [example](https://github.com/dcos/dcos-mesos-modules/compare/f6fa27d7c40f4207ba3bb2274e2cfe79b62a395a...6660b90fbbf69a15ef46d0184e36755881d6a5ae)
  - [ ] Test Results: [link to CI job test results for component]
  - [ ] Code Coverage (if available): [link to code coverage report]

### CODE REVIEW

**Comment by Gilbert88** [comment]
@mesosphere-mergebot bump-ee

**Comment by Gilbert88** [comment]
@vinodkone @adam-mesos @qianzhangxa @jieyu Could you please help review this PR?

File: packages/mesos/build [hunk 1]
```diff
@@ -9,10 +9,12 @@ mkdir -p build
 pushd build
 # TODO(cmaloney): --with-glog=/usr --with-protobuf=/usr --with-boost=/usr
 # TODO(cmaloney): DESTDIR builds so we don't have to build as root?
+# TODO(nfnt): --with-grpc=/usr
 LDFLAGS="-Wl,-rpath,/opt/mesosphere/lib" "/pkg/src/mesos/configure" \
   --prefix="$PKG_PATH" --enable-optimize --disable-python \
   --enable-libevent --enable-ssl \
   --enable-install-module-dependencies \
+  --enable-grpc \
```
Comment (vinodkone) [inline packages/mesos/build]: this seems unrelated to this change?

**Comment by Gilbert88** [comment]
@vinodkone Yes, tested on a real dc/os cluster. Verified that unused container image layers are gc-ed correctly.

**Comment by qianzhangxa** [comment]
@mesosphere-mergebot label Ship It

File: gen/dcos-config.yaml [hunk 2]
```diff
@@ -367,22 +367,31 @@ package:
...
-                "nameservers": [ 
+                "nameservers": [
                                 "198.51.100.1",
                                 "fd01:d::c633:6401"
                 ]
               }
             }
         ]
       }
+  - path: /etc/mesos-slave-image-gc-config.json
```
Comment (adam-mesos) [inline gen/dcos-config.yaml]: s/slave/agent/g ?

Comment (Gilbert88) [reply gen/dcos-config.yaml]: Rebased. Thanks, @adam-mesos !

About the renaming, we may not want to do that now. Because we may need to name all other agent related config files as `mesos-agent-xxxxx`, for instance, now we have:
/opt/mesosphere/etc/mesos-slave
/opt/mesosphere/etc/mesos-slave-public
/opt/mesosphere/etc/mesos-slave-dns.json
/opt/mesosphere/etc/mesos-slave-common
/opt/mesosphere/etc/mesos-slave-modules
/opt/mesosphere/etc/mesos-slave-public

**Comment by Gilbert88** [comment]
@mesosphere-mergebot bump-ee

File: gen/dcos-config.yaml [hunk 3]
```diff
@@ -367,22 +367,31 @@ package:
...
+    content: |
+      {
+        "image_disk_headroom": 0.1,
+        "image_disk_watch_interval": {
+          "nanoseconds": 300000000000
```
Comment (adam-mesos) [inline gen/dcos-config.yaml]: Why is this in ns when the default value is going to be so large? Wouldn't "seconds" have been sufficient?

File: gen/dcos-config.yaml [hunk 4]
```diff
@@ -367,22 +367,31 @@ package:
...
+        },
+        "excluded_images": []
```
Comment (adam-mesos) [inline gen/dcos-config.yaml]: Does this need to be included if it's an empty array?
Or are you just suggesting where an admin could add images to exclude?

File: gen/dcos-config.yaml (hunk 3 above)
Comment (Gilbert88) [reply gen/dcos-config.yaml]: This was discussed in the protobuf design patch: https://reviews.apache.org/r/64265/#comment270885

tl;dr: Due to the tech debt, we don't have a sufficient way to represent time duration in protobuf.

From operator's perspective, `nanoseconds` is inconvenient but accurate.

File: gen/dcos-config.yaml (hunk 4 above)
Comment (Gilbert88) [reply gen/dcos-config.yaml]: It is an optional field and is not necessary to be included.

But I decided to pass in an empty array to let dc/os operators know there is such a field in container image gc config file and they can consume it if they did not read the image gc documents. Some operators might be very interested in using `excluded_images` (e.g., Uber's use cases).

**Comment by adam-mesos** [comment]
@vinodkone Can you re-review this PR (and the EE bump PR) after the rebase, so your approval counts again (and in both places)?

**Comment by vinodkone** [comment]
@adam-mesos  Approved. Seems weird that I need to re-approve after every rebase? Can't we trust the contributor?

**Comment by adam-mesos** [comment]
@vinodkone Mergebot doesn't distinguish rebases from other SHA changes (yet?), and you've only approved the previous SHA. It's hard for a bot to know that other changes from master (e.g. a Mesos bump) didn't affect the validity of your approval, or that additional commits added to this PR don't require review.
